interpreter/interpreter.go

```diff
@@ -151,6 +156,229 @@ func CompileRegexConstants(regexOptimizations ...*RegexOptimization) Interpretab
 	return decRegexOptimizer(regexOptimizations...)
 }
 
+type typeVerifier func(int64, ...*types.TypeValue) (bool, error)
+
+// CheckStringFormat checks the syntax and cardinality of any string.format calls present in the expression and reports
+// any errors at compile time.
+func CheckStringFormat(enableCheckArgTypes bool, verifier typeVerifier) InterpretableDecorator {
+	return func(inter Interpretable) (Interpretable, error) {
+		call, ok := inter.(InterpretableCall)
+		if !ok {
+			return inter, nil
+		}
+		if call.OverloadID() != "string_format" {
+			return inter, nil
+		}
+		args := call.Args()
+		if len(args) != 2 {
+			return nil, fmt.Errorf("wrong number of arguments to string.format (expected 2, got %d)", len(args))
+		}
+		fmtStrInter, ok := args[0].(InterpretableConst)
+		if !ok {
+			return inter, nil
+		}
+		var fmtArgsInter InterpretableConstructor
+		fmtArgsInter, ok = args[1].(InterpretableConstructor)
+		if !ok {
+			// check for an evalWatch wrapper, since evalWatch implements
+			// Interpretable but not InterpretableConstructor (but may be wrapping
+			// the InterpretableConstructor we're looking for)
+			ew, ok := args[1].(*evalWatch)
+			if !ok {
+				return inter, nil
+			}
+			fmtArgsInter, ok = ew.Interpretable.(InterpretableConstructor)
+			if !ok {
+				return inter, nil
+			}
+		}
+		if fmtArgsInter.Type() != types.ListType {
+			// don't necessarily return an error since the list may be DynType
+			return inter, nil
+		}
+		formatStr := fmtStrInter.Value().Value().(string)
+		formatArgs := fmtArgsInter.InitVals()
+
+		formatCallback := &formatCheck{
+			formatArgs:          formatArgs,
+			enableCheckArgTypes: enableCheckArgTypes,
+			verifier:            verifier,
+		}
+		// use a placeholder locale, since locale doesn't affect syntax
+		_, err := functions.ParseFormatString(formatStr, formatCallback, "en_US")
+		if err != nil {
+			return nil, err
+		}
+		seenArgs := formatCallback.argsRequested
+		if len(formatArgs) > seenArgs {
+			return nil, fmt.Errorf("too many arguments supplied to string.format (expected %d, got %d)", seenArgs, len(formatArgs))
+		}
```

Comment:
Are excess args ignored? Is that documented?

Reply:
If I'm understanding you correctly (is something like `"%d %d".format([1, 2, 3])` supported), that's what this case is meant to handle - `len(formatArgs)` is `len([1, 2, 3])` in that example. If there were more formatting clauses (say, 4 total uses of `%d` in that example), then that would be caught in `parseFormatArg`.

It is true that this specific `len` check is not done if compile-time checking is disabled/unavailable - I'll document that.

Reply:
Gotcha. Should it be `seenClauses` ? I'm seeing a mix of the terms "args" and "clauses" and it can be tedious trying to untangle which means what.

Reply:
You're right, that mix is confusing. I'll rename and do a pass on the PR to see if there's any similar refactors I should perform.

Reply:
(marking as resolved, didn't notice anything else)